dcos/cosmos.py

```diff
@@ -0,0 +1,401 @@
+            remove_nones(json),
+            **kwargs)
+
+    def _call_cosmos(self,
+                     url,
+                     http_request_type,
+                     headers_preference,
+                     data=None,
+                     json=None,
+                     **kwargs):
+        """
+        Gets a Response object obtained by calling cosmos
+        at the url 'url'. Will attempt each of the headers
+        in headers_preference in order until success.
+
+        :param url: the url of a cosmos endpoint
+        :type url: str
+        :param headers_preference: a list of request headers
+        in order of preference. Each header
+        will be attempted until they all fail or the request succeeds.
+        :type headers_preference: list[dict[str, str]]
+        :param data: the request's body
+        :type data: dict | bytes | file-like object
+        :param json: JSON request body
+        :type json: dict
+        :param kwargs: Additional arguments to requests.request
+                       (see py:func:`request`)
+        :type kwargs: dict
+        :return: response returned by calling cosmos at url
+        :rtype: requests.Response
+        """
+        try:
+            headers = headers_preference[0]
+            if http_request_type is 'post':
+                response = http.post(
+                    url, data=data, json=json, headers=headers, **kwargs)
+            else:
+                response = http.get(
+                    url, data=data, json=json, headers=headers, **kwargs)
+            if not _matches_expected_response_header(headers,
+                                                     response.headers):
+                raise DCOSException(
+                    'Server returned incorrect response type, '
+                    'expected {} but got {}'.format(
+                        headers.get('Accept'),
+                        response.headers.get('Content-Type')))
+        except DCOSBadRequest as e:
+            if len(headers_preference) > 1:
+                # reattempt with one less item in headers_preference
+                response = self._call_cosmos(url,
+                                             http_request_type,
+                                             headers_preference[1:],
+                                             data,
+                                             json,
+                                             **kwargs)
+            else:
+                raise e
+        return response
+
+    def _get_endpoint_url(self, endpoint):
+        """
+        Gets the url for the cosmos endpoint 'endpoint'
+
+        :param endpoint: a cosmos endpoint, of the form 'x/y',
+        for example 'package/repo/add' or 'service/start'
+        :type endpoint: str
+        :return: the url of endpoint
+        :rtype: str
+        """
+        return urllib.parse.urljoin(self.cosmos_url, endpoint)
+
+    def _get_request_version_preferences(self, endpoint):
```

Comment:
the method name is the same length as the code in this method :) We should just call `self._request_versions.get(endpoin`